### PR TITLE
State: Migrate lib/site-roles to Redux state

### LINKED ISSUE #7921: State: Migrate lib/site-roles to Redux state
Related: #5046

The [`lib/site-roles` Flux store](https://github.com/Automattic/wp-calypso/tree/master/client/lib/site-roles) is a legacy store which uses deprecated data patterns. It should be updated to align with our [current data recommendations](https://github.com/Automattic/wp-calypso/blob/master/docs/our-approach-to-data.md#current-recommendations). Specifically, this may include implementing all or some of the following Redux state behaviors:
- [ ] Reducer
- [ ] Action creators
- [ ] Selectors
- [ ] Query components

The implementation of each of these behaviors should be accompanied by tests.

When submitting a pull request, reference this issue and update the checklist above to reflect progress.

### PR DESCRIPTION
**Pull request by tyxla**
This PR migrates from the legacy [`lib/site-roles` Flux store](https://github.com/Automattic/wp-calypso/tree/master/client/lib/site-roles) to Redux - fixes #7921. The PR takes care of the following:

* Creates Redux actions, reducer, schema, selectors and README for site roles.
* Adds tests for the above.
* Introduces a `QuerySiteRoles` query component.
* Updates the `RoleSelect` component to use Redux and converts it to a stateless functional component.
* Removes the legacy `lib/site-roles` Flux store.

To test:

1. Checkout this branch.
2. Use one of your **public** WordPress.com sites for the next steps.
3. Go to **People**, click on a user and verify the roles load properly in the **Role** field.
4. Click on the button to add a new user and verify the roles load properly in the **Role** field.
5. Verify step 3 and 4 work properly with a clean Redux state.
6. Verify the last role in the role select field is called `Follower`.
7. Use one of your **private** WordPress.com sites, go through steps 3 to 6, but this time verify the last role is called "Viewer".
8. Use one of your Jetpack sites and go through step 3 to 5, verify all works as expected.
9. Verify all tests pass: `npm run test-client client/state/site-roles/`

### CODE REVIEW

File: client/lib/wpcom-undocumented/lib/site.js [hunk 1]
```diff
@@ -118,7 +118,7 @@ UndocumentedSite.prototype.shortcodes = function( attributes, callback ) {
 };
 
 UndocumentedSite.prototype.getRoles = function( callback ) {
-	this.wpcom.withLocale().req.get( '/sites/' + this._id + '/roles', {}, callback );
+	return this.wpcom.withLocale().req.get( '/sites/' + this._id + '/roles', {}, callback );
```
Comment (aduth) [inline client/lib/wpcom-undocumented/lib/site.js]: If you rebase, this change will be omitted now (#9116), though merge will figure it out too in any case.

File: client/my-sites/people/role-select/index.jsx [hunk 2]
```diff
@@ -2,138 +2,87 @@
...
+
+	return {
+		display_name: displayName,
+		name: 'follower'
+	};
+};
+
+const RoleSelect = ( props ) => {
+	const { siteRoles, site, includeFollower, siteId, id, explanation, translate } = props;
+	const roles = siteRoles && siteRoles.slice( 0 );
```
Comment (aduth) [inline client/my-sites/people/role-select/index.jsx]: Instead of `Array#slice` to copy the array because we use `Array#push` below, we could instead use `Array#concat`:

```js
let { siteRoles } = props;
if ( siteRoles && includeFollower ) {
	siteRoles = siteRoles.concat( getWPCOMFollowerRole( props ) );
}
```

File: client/my-sites/people/role-select/index.jsx [hunk 3]
```diff
@@ -2,138 +2,87 @@
...
-			return;
-		}
-
-		debug( 'Fetching roles for ' + siteId );
+import QuerySites from 'components/data/query-sites';
+import QuerySiteRoles from 'components/data/query-site-roles';
+import { getSite } from 'state/sites/selectors';
+import { getSiteRoles } from 'state/site-roles/selectors';
+
+const getWPCOMFollowerRole = ( { site, translate } ) => {
```
Comment (aduth) [inline client/my-sites/people/role-select/index.jsx]: Per [naming conventions on abbreviations](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#naming-conventions) this should probably be `getWpcomFollowerRole`

File: client/my-sites/people/role-select/index.jsx [hunk 4]
```diff
@@ -2,138 +2,87 @@
...
+	}
 
-		// defer fetch requests to avoid dispatcher conflicts
-		setTimeout( function() {
-			const fetching = RolesStore.isFetching( siteId );
-			if ( fetching ) {
-				return;
+	return (
+		<FormFieldset key={ siteId } disabled={ ! roles }>
+			{ siteId &&
```
Comment (aduth) [inline client/my-sites/people/role-select/index.jsx]: This won't work the way you think it does. The logic is such that it'll check the truthiness of each and stop at whichever is falsey or the final value, so it'd either render `null` for an empty site ID or the `<QuerySiteRoles />` component, but never the `<QuerySites />` component.

Instead, you might want:

```jsx
{ siteId && <QuerySites siteId={ siteId } /> }
{ siteId && <QuerySiteRoles siteId={ siteId } /> }
```

Or:

```jsx
{ siteId && [
	<QuerySites key="query-sites" siteId={ siteId } />,
	<QuerySiteRoles key="query-site-roles" siteId={ siteId } />
] }
```

File: client/my-sites/people/role-select/index.jsx [hunk 5]
```diff
@@ -2,138 +2,87 @@
...
-						} )
-					}
-				</FormSelect>
-				{ this.props.explanation &&
-					<FormSettingExplanation>
-						{ this.props.explanation }
-					</FormSettingExplanation>
+			<FormLabel htmlFor={ id }>
+				{ translate( 'Role', {
+					context: 'Text that is displayed in a label of a form.'
```
Comment (aduth) [inline client/my-sites/people/role-select/index.jsx]: Acknowledging it was there previously, I find this to be a strange `context` for the string. Why do we need context in this case? Role doesn't seem specific to a form.

File: client/my-sites/people/role-select/index.jsx [hunk 6]
```diff
@@ -2,138 +2,87 @@
...
+				} ) }
+			</FormLabel>
+			<FormSelect { ...omit( props, omitProps ) }>
+				{
+					roles && roles.map( ( role ) => {
```
Comment (aduth) [inline client/my-sites/people/role-select/index.jsx]: The nice thing about Lodash's `map` is that it would handle the empty `roles` for you automatically.

[It's also generally faster](https://jsperf.com/js-map-vs-lodash-map).

I don't feel particularly strongly about changing this.

File: client/state/site-roles/actions.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+			dispatch( {
+				type: SITE_ROLES_RECEIVE,
+				siteId,
+				roles
+			} );
+		} ).catch( ( error ) => {
+			dispatch( {
+				type: SITE_ROLES_REQUEST_FAILURE,
+				siteId,
+				error: pick( error, [ 'error', 'status', 'message' ] )
```
Comment (aduth) [inline client/state/site-roles/actions.js]: That's a lot of work to shape an object that we don't use 😄

File: client/lib/wpcom-undocumented/lib/site.js (hunk 1 above)
Comment (tyxla) [inline client/lib/wpcom-undocumented/lib/site.js]: :+1: true - I rebased, and it did go away. Great job with #9116 btw.

File: client/my-sites/people/role-select/index.jsx (hunk 5 above)
Comment (tyxla) [inline client/my-sites/people/role-select/index.jsx]: Makes sense - I don't think this would help the translators in any way, so removing that.

File: client/state/site-roles/actions.js (hunk 7 above)
Comment (tyxla) [inline client/state/site-roles/actions.js]: :+1: Removing this, since we don't use the `error` at all.

**Comment by tyxla** [comment]
Thanks for the review, @aduth, the PR is ready for a new one.

File: client/my-sites/people/role-select/index.jsx (hunk 6 above)
Comment (aduth) [inline client/my-sites/people/role-select/index.jsx]: With latest changes, you don't need the `siteRoles &&` check anymore. `map` will return an empty array if passed a falsey value.

>Oh, I was left with the impression that the native `Array.map` would be faster.

Lodash "cheats" and sometimes forgoes full spec compliancy vs. native counterparts in favor of performance. In the case of array iteration ignoring [sparse array checks](http://stackoverflow.com/a/18884620).

See also: 

- https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore/issues/17
- https://www.youtube.com/watch?v=OP8jdbcDfaA

File: client/my-sites/people/role-select/index.jsx [hunk 8]
```diff
@@ -2,138 +2,83 @@
...
+		'explanation',
+		'siteRoles',
+		'dispatch',
+		'moment',
+		'numberFormat',
+		'translate'
+	];
+
+	if ( site && siteRoles && includeFollower ) {
+		siteRoles = siteRoles.concat( getWpcomFollowerRole( props ) );
```
Comment (aduth) [inline client/my-sites/people/role-select/index.jsx]: Do you think this is something that belongs in a state selector? In other words, should `getSiteRoles` include the follower role automatically when the site is not private?

File: client/my-sites/people/role-select/index.jsx [hunk 9]
```diff
@@ -2,138 +2,83 @@
...
-						{ this.props.explanation }
-					</FormSettingExplanation>
+	return (
+		<FormFieldset key={ siteId } disabled={ ! siteRoles }>
+			{ siteId && <QuerySites siteId={ siteId } /> }
+			{ siteId && <QuerySiteRoles siteId={ siteId } /> }
+			<FormLabel htmlFor={ id }>
+				{ translate( 'Role' ) }
+			</FormLabel>
+			<FormSelect { ...omit( props, omitProps ) }>
```
Comment (aduth) [inline client/my-sites/people/role-select/index.jsx]: If we could figure out based on usage what's expected for this component to pass through (e.g. `onChange`), would be nice to avoid the `omit`. I see in a few cases there's an `id` and `name` passed, though I'm doubtful those are really necessary.

Comment (tyxla) [inline client/my-sites/people/role-select/index.jsx]: Good call. The list of `omitProps` has grown. Let's keep this for another PR.
